File: python/src/dashgps/writers/summaryw.py
"""Summary JSON. spec/21-outputs.md.

A file that failed to parse still gets an entry, with `error` set. A batch never silently drops
a file.
"""

from ..fmt import P_DIST, P_DUR, P_LATLON, byte_key, fixed, iso_local, json_str


def _num(v, prec):
    s = fixed(v, prec)
    return s if s else "null"


def write(out, entries, groups, ctx):
    entries = sorted(entries, key=lambda e: byte_key(e["name"]))
    out.append("{\n")
    out.append('  "generator": ' + json_str("dashgps " + ctx.get("version", "0.0.0")) + ",\n")
    out.append('  "time_is_naive": ' + ("true" if ctx.get("time_is_naive", True) else "false")
               + ",\n")
    out.append('  "files": [\n')
    for i, e in enumerate(entries):
        out.append("    {\n")
        out.append('      "name": ' + json_str(e["name"]) + ",\n")
        out.append('      "size": ' + str(e["size"]) + ",\n")
        out.append('      "format": ' + (json_str(e["format"]) if e["format"] else "null") + ",\n")
        out.append('      "status": ' + (json_str(e["status"]) if e["status"] else "null") + ",\n")
        out.append('      "records": ' + str(e["records"]) + ",\n")
        out.append('      "points": ' + str(e["points"]) + ",\n")
        out.append('      "dropped_nofix": ' + str(e["dropped_nofix"]) + ",\n")
        out.append('      "t_start": ' + (json_str(e["t_start"]) if e["t_start"] else "null")
                   + ",\n")
        out.append('      "t_end": ' + (json_str(e["t_end"]) if e["t_end"] else "null") + ",\n")
        bb = e["bbox"]
        if bb is None:
            out.append('      "bbox": null,\n')
        else:
            out.append('      "bbox": [' + ", ".join(fixed(v, P_LATLON) for v in bb) + "],\n")
        out.append('      "warnings": [' + ", ".join(json_str(w) for w in e["warnings"]) + "],\n")
        out.append('      "error": ' + (json_str(e["error"]) if e["error"] else "null") + "\n")
        out.append("    }")
        out.append(",\n" if i < len(entries) - 1 else "\n")
    out.append("  ],\n")

    out.append('  "groups": [\n')
    for i, g in enumerate(groups):
        kept = [r for r in g.runs if not r.glitch]
        dist = 0.0
        for r in kept:
            dist += r.distance_km
        glitch_pts = 0
        for p in g.points:
            if p.outlier:
                glitch_pts += 1
        dur = (g.points[-1].t - g.points[0].t) if g.points else 0.0
        out.append("    {\n")
        out.append('      "label": ' + json_str(g.label) + ",\n")
        out.append('      "points": ' + str(len(g.points)) + ",\n")
        out.append('      "outlier_points": ' + str(glitch_pts) + ",\n")
        out.append('      "dropped_duplicate_times": ' + str(g.dropped_dupe) + ",\n")
        out.append('      "runs": ' + str(len(g.runs)) + ",\n")
        out.append('      "glitch_runs": ' + str(len(g.runs) - len(kept)) + ",\n")
        out.append('      "distance_km": ' + _num(dist, P_DIST) + ",\n")
        out.append('      "duration_s": ' + _num(dur, P_DUR) + ",\n")
        out.append('      "start": ' + (json_str(iso_local(g.points[0].t)) if g.points else "null")
                   + ",\n")
        out.append('      "end": ' + (json_str(iso_local(g.points[-1].t)) if g.points else "null")
                   + ",\n")
        out.append('      "sources": [' + ", ".join(json_str(s) for s in g.sources) + "]\n")
        out.append("    }")
        out.append(",\n" if i < len(groups) - 1 else "\n")
    out.append("  ],\n")

    tot_pts = 0
    tot_dist = 0.0
    for g in groups:
        tot_pts += len(g.points)
        for r in g.runs:
            if not r.glitch:
                tot_dist += r.distance_km
    ok = 0
    for e in entries:
        if not e["error"]:
            ok += 1
    out.append('  "totals": {\n')
    out.append('    "files": ' + str(len(entries)) + ",\n")
    out.append('    "files_parsed": ' + str(ok) + ",\n")
    out.append('    "groups": ' + str(len(groups)) + ",\n")
    out.append('    "points": ' + str(tot_pts) + ",\n")
    out.append('    "distance_km": ' + _num(tot_dist, P_DIST) + "\n")
    out.append("  }\n}\n")

File: js/src/writers/summaryw.js
// Summary JSON. spec/21-outputs.md  (mirrors python/src/dashgps/writers/summaryw.py)
//
// A file that failed to parse still gets an entry, with `error` set. A batch never silently
// drops a file.

import { P_DIST, P_DUR, P_LATLON, cmpNames, fixed, isoLocal, jsonStr } from "../fmt.js";

function num(v, prec) {
  const s = fixed(v, prec);
  return s || "null";
}

export function write(out, entries, groups, ctx) {
  entries = entries.slice().sort((a, b) => cmpNames(a.name, b.name));
  out.push("{\n");
  out.push('  "generator": ' + jsonStr("dashgps " + (ctx.version || "0.0.0")) + ",\n");
  out.push('  "time_is_naive": ' + (ctx.timeIsNaive !== false ? "true" : "false") + ",\n");
  out.push('  "files": [\n');
  for (let i = 0; i < entries.length; i++) {
    const e = entries[i];
    out.push("    {\n");
    out.push('      "name": ' + jsonStr(e.name) + ",\n");
    out.push('      "size": ' + e.size + ",\n");
    out.push('      "format": ' + (e.format ? jsonStr(e.format) : "null") + ",\n");
    out.push('      "status": ' + (e.status ? jsonStr(e.status) : "null") + ",\n");
    out.push('      "records": ' + e.records + ",\n");
    out.push('      "points": ' + e.points + ",\n");
    out.push('      "dropped_nofix": ' + e.droppedNofix + ",\n");
    out.push('      "t_start": ' + (e.tStart ? jsonStr(e.tStart) : "null") + ",\n");
    out.push('      "t_end": ' + (e.tEnd ? jsonStr(e.tEnd) : "null") + ",\n");
    if (!e.bbox) out.push('      "bbox": null,\n');
    else out.push('      "bbox": [' + e.bbox.map((v) => fixed(v, P_LATLON)).join(", ") + "],\n");
    out.push('      "warnings": [' + e.warnings.map(jsonStr).join(", ") + "],\n");
    out.push('      "error": ' + (e.error ? jsonStr(e.error) : "null") + "\n");
    out.push("    }");
    out.push(i < entries.length - 1 ? ",\n" : "\n");
  }
  out.push("  ],\n");

  out.push('  "groups": [\n');
  for (let i = 0; i < groups.length; i++) {
    const g = groups[i];
    const kept = g.runs.filter((r) => !r.glitch);
    let dist = 0;
    for (const r of kept) dist += r.distanceKm;
    let glitchPts = 0;
    for (const p of g.points) if (p.outlier) glitchPts += 1;
    const dur = g.points.length ? g.points[g.points.length - 1].t - g.points[0].t : 0;
    out.push("    {\n");
    out.push('      "label": ' + jsonStr(g.label) + ",\n");
    out.push('      "points": ' + g.points.length + ",\n");
    out.push('      "outlier_points": ' + glitchPts + ",\n");
    out.push('      "dropped_duplicate_times": ' + (g.droppedDupe || 0) + ",\n");
    out.push('      "runs": ' + g.runs.length + ",\n");
    out.push('      "glitch_runs": ' + (g.runs.length - kept.length) + ",\n");
    out.push('      "distance_km": ' + num(dist, P_DIST) + ",\n");
    out.push('      "duration_s": ' + num(dur, P_DUR) + ",\n");
    out.push('      "start": ' + (g.points.length ? jsonStr(isoLocal(g.points[0].t)) : "null") + ",\n");
    out.push('      "end": ' + (g.points.length
      ? jsonStr(isoLocal(g.points[g.points.length - 1].t)) : "null") + ",\n");
    out.push('      "sources": [' + g.sources.map(jsonStr).join(", ") + "]\n");
    out.push("    }");
    out.push(i < groups.length - 1 ? ",\n" : "\n");
  }
  out.push("  ],\n");

  let totPts = 0;
  let totDist = 0;
  for (const g of groups) {
    totPts += g.points.length;
    for (const r of g.runs) if (!r.glitch) totDist += r.distanceKm;
  }
  let ok = 0;
  for (const e of entries) if (!e.error) ok += 1;
  out.push('  "totals": {\n');
  out.push('    "files": ' + entries.length + ",\n");
  out.push('    "files_parsed": ' + ok + ",\n");
  out.push('    "groups": ' + groups.length + ",\n");
  out.push('    "points": ' + totPts + ",\n");
  out.push('    "distance_km": ' + num(totDist, P_DIST) + "\n");
  out.push("  }\n}\n");
}
